connectors/src/connectors/notion/temporal/activities.ts

```diff
@@ -2310,3 +2360,95 @@ async function renderPageSection({
 function redisGarbageCollectorKey(connectorId: ModelId): string {
   return `notion-garbage-collector-${connectorId}`;
 }
+
+export async function databaseHasStructuredDataEnabled({
+  databaseId,
+  connectorId,
+}: {
+  databaseId: string;
+  connectorId: ModelId;
+}): Promise<boolean> {
+  const dbModel = await NotionDatabase.findOne({
+    where: {
+      connectorId,
+      notionDatabaseId: databaseId,
+    },
+  });
+
+  return !!dbModel?.structuredDataEnabled;
+}
+
+export async function upsertDatabaseStructuredDataFromCache({
+  databaseId,
+  connectorId,
+  topLevelWorkflowId,
+  loggerArgs,
+}: {
+  databaseId: string;
+  connectorId: number;
+  topLevelWorkflowId: string;
+  loggerArgs: Record<string, string | number>;
+}): Promise<void> {
+  const connector = await Connector.findOne({
+    where: {
+      type: "notion",
+      id: connectorId,
+    },
+  });
+
+  if (!connector) {
+    throw new Error("Could not find connector");
+  }
+
+  const localLogger = logger.child({
+    ...loggerArgs,
+    workspaceId: connector.workspaceId,
+    dataSourceName: connector.dataSourceName,
+    databaseId,
+  });
+
+  const dbModel = await NotionDatabase.findOne({
+    where: {
+      connectorId,
+      notionDatabaseId: databaseId,
+    },
+  });
+
+  if (!dbModel?.structuredDataEnabled) {
+    localLogger.info("Structured data not enabled for database (skipping).");
+    return;
+  }
+
+  const pageCacheEntries = await NotionConnectorPageCacheEntry.findAll({
```

Comment:
We have this because notion is split in many sub-workflows that "cache" pages and databases. At the end there is an activity that performs upserts from this "cache" table. This is needed because upserting a single page from notion could result in hundreds of API calls and you never know how many ahead of time (pages have blocks, blocks have child blocks etc...). Pages can also have child databases which we want to render inline (so we need to have the child pages in this "cache")